File: prime/src/main/java/org/ostelco/prime/events/EventListeners.java
package org.ostelco.prime.events;

import org.ostelco.prime.ocs.OcsState;
import org.ostelco.prime.storage.ProductCatalogItem;
import org.ostelco.prime.storage.ProductDescriptionCacheImpl;
import org.ostelco.prime.storage.PurchaseRequestListener;
import org.ostelco.prime.storage.StorageInitiatedEventExecutor;
import org.ostelco.prime.storage.entities.PurchaseRequestImpl;
import org.ostelco.prime.storage.entities.Subscriber;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.Collection;

import static com.google.common.base.Preconditions.checkNotNull;

// Badly named class
public final class EventListeners {

    private static final Logger LOG = LoggerFactory.getLogger(EventListeners.class);

    private final StorageInitiatedEventExecutor executor;

    private final OcsState ocsState;

    public EventListeners(final OcsState ocsState) {
        this.executor = new StorageInitiatedEventExecutor();
        this.ocsState = checkNotNull(ocsState);
    }

    public void loadSubscriberBalanceDataFromFirebaseToInMemoryStructure(
            final Collection<Subscriber> subscribers) {
        LOG.info("Loading initial balance from storage to in-memory OcsState");
        for (final Subscriber subscriber : subscribers) {
            ocsState.injectSubscriberIntoOCS(subscriber);
        }
    }

    private long getMillisSinceEpoch() {
        return Instant.now().toEpochMilli();
    }


    public Void purchaseRequestListener(final String key, final PurchaseRequestImpl req) {
        req.setId(key);
        req.setMillisSinceEpoch(getMillisSinceEpoch());
        executor.onPurchaseRequest(req);
        return null; // XXX Hack to satisfy BiFunction's void return type
    }

    public void productCatalogItemListener(final ProductCatalogItem item) {
        ProductDescriptionCacheImpl.
                getInstance(). // XXX This is an awful hack!
                addTopupProduct(item.getSku(), item.getNoOfBytes());
    }

    // XXX I don't like this!
    public void addPurchaseRequestListener(final PurchaseRequestListener listener) {
        executor.addPurchaseRequestListener(listener);
    }
}

File: prime/src/main/kotlin/org/ostelco/prime/events/EventListeners.kt
package org.ostelco.prime.events

import com.google.common.base.Preconditions.checkNotNull
import org.ostelco.prime.logger
import org.ostelco.prime.ocs.OcsState
import org.ostelco.prime.storage.ProductCatalogItem
import org.ostelco.prime.storage.ProductDescriptionCacheImpl
import org.ostelco.prime.storage.PurchaseRequestHandler
import org.ostelco.prime.storage.StorageInitiatedEventExecutor
import org.ostelco.prime.storage.entities.PurchaseRequestImpl
import org.ostelco.prime.storage.entities.Subscriber
import java.time.Instant

// Badly named class
class EventListeners(ocsState: OcsState) {

    private val LOG by logger()

    private val executor: StorageInitiatedEventExecutor = StorageInitiatedEventExecutor()

    private val ocsState: OcsState = checkNotNull(ocsState)

    private val millisSinceEpoch: Long
        get() = Instant.now().toEpochMilli()

    fun loadSubscriberBalanceDataFromFirebaseToInMemoryStructure(
            subscribers: Collection<Subscriber>) {
        LOG.info("Loading initial balance from storage to in-memory OcsState")
        for (subscriber in subscribers) {
            ocsState.injectSubscriberIntoOCS(subscriber)
        }
    }


    fun purchaseRequestListener(key: String, req: PurchaseRequestImpl) {
        req.setId(key)
        req.setMillisSinceEpoch(millisSinceEpoch)
        executor.onPurchaseRequest(req)
        // return null // XXX Hack to satisfy BiFunction's void return type
    }

    fun productCatalogItemListener(item: ProductCatalogItem) {
        val sku = item.sku
        val noOfBytes = item.noOfBytes
        if (sku != null && noOfBytes != null) {
            ProductDescriptionCacheImpl. // XXX This is an awful hack!
                    addTopupProduct(sku, noOfBytes)
        }
    }

    // XXX I don't like this!
    fun addPurchaseRequestHandler(handler: PurchaseRequestHandler) {
        executor.addPurchaseRequestHandler(handler)
    }
}
